pkg/watermark/fetch/edge_fetcher.go

```diff
@@ -205,16 +210,13 @@ func (e *edgeFetcher) GetHeadWMB(fromPartitionIdx int32) wmb.WMB {
 	defer e.Unlock()
 	// update the last processed watermark for the partition
 	e.lastProcessedWm[fromPartitionIdx] = headWMB.Watermark
-	// we only consider idle watermark if it is smaller than or equal to min of all the last processed watermarks.
```

Comment:
Why is this removed?

Reply:
this check is moved into EdgeFetcherSet::GetHeadWMB(). The check needs to be made across all of the Edges.

Reply:
Do we only call `EdgeFetcher::GetHeadWMB()` in `EdgeFetcherSet::GetHeadWMB()`?? Removing this part will result in the `EdgeFetcher::GetHeadWMB()` returning wrong value.

Reply:
yes. For the purpose of `EdgeFetcherSet::GetHeadWMB()` we can't call it here. If you have a concern about somebody calling this directly in the future as far as this check being removed let me know.

Reply:
Could you add a TODO or comment (regarding `this check is moved into EdgeFetcherSet::GetHeadWMB() The check needs to be made across all of the Edges` and `Removing this part will result in the EdgeFetcher::GetHeadWMB() returning a wrong value`)  to inform that this could have a potential problem, and also explain that we have the check in EdgeFetcherSet?

Reply:
Conceivably the note could say that it's the responsibility of the caller to verify it?

Reply:
Or do you feel like the contract of the method says that it will definitively do this?

Reply:
> the contract of the method says that it will definitively do this

We have the `GetHeadWMB()` in the interface. The two datatypes that implement this interface have different behavior: EdgeFetcher doesn't compare the WMB with lastProcessedWMs while EdgeFetcherSet does. This is inconsistent and may confuse developers.

It's up to us how to define what a method does. I think either having the comparison or not having it works as long as the datatypes that implement this method in a way what the method is designed to be.

Reply:
I agree that consistency is good.

Reply:
I think I'm going to move the check into the callers and this method will make no promises as far as that.

Reply:
Do you consider it leaking if the caller does a sanity check to verify that the Idle Watermark is not more advanced than `GetWatermark()`?

Reply:
Perhaps I should also revisit whether this definitely needs to be yanked out here. At the time, I decided it did but maybe that decision needs revisiting.